### PR TITLE
[AutoParallel-PIR] AutoParallel Main Framework for PIR mode

### PR DESCRIPTION
**Pull request by JZ-LIANG**
### PR types
New features

### PR changes
Others

### Description

Pcard-67164

Organize the main logic of Auto Parallel in PIR mode. Its logic consist of Five parts:

1. **Program Completion:** build a completed program with forward-backward-optimizer from a forward program. 
2. **Parallelism Completion:** rule-based entire-graph sharding propagation(Semi-Auto) Or algorithm/random-based parallel search(Fully-Auto).
3. **Graph partition:** Partition(Pipeline-like parallel) and Reshard Pass(SPMD parallel).
4. **Parallel related Optimization Pass**. 
5. **Communicator Maintaining**

### CODE REVIEW

**Review by hitywt — APPROVED** [review]
LGTM

File: paddle/fluid/pir/dialect/distributed/transforms/mix_to_dist_pass.cc [hunk 1]
```diff
@@ -134,15 +120,13 @@ void VerifyBlock(pir::Block* block) {
...
-    VLOG(0) << "verifying op name " << op_item->name();
   }
 }
 
 std::shared_ptr<pir::Program> MixToDistPass(pir::Program* prog) {
-  // if (FLAGS_print_ir) {
-  std::cout << "IR before MixToDist Pass = " << *prog << std::endl;
-  // }
+  if (FLAGS_print_ir) {
+    std::cout << "IR before MixToDist Pass = " << *prog << std::endl;
```
Comment (winter-wang) [inline paddle/fluid/pir/dialect/distributed/transforms/mix_to_dist_pass.cc:128]: 目前好像不允许使用std::cout, 使用VLOG来打印日志。

**Review by winter-wang — APPROVED** [review]
LGTM

File: paddle/fluid/pir/dialect/distributed/transforms/mix_to_dist_pass.cc (hunk 1 above)
Comment (JZ-LIANG) [reply paddle/fluid/pir/dialect/distributed/transforms/mix_to_dist_pass.cc:128]: fix in next pr

**Review by zhangbo9674 — APPROVED** [review]

**Review by zhiqiu — APPROVED** [review]
LGTM

**Pull request merged**